src/FastTags/LexerM.hs

```diff
@@ -0,0 +1,780 @@
+----------------------------------------------------------------------------
+-- |
+-- Module      :  FastTags.LexerTypes
+-- Copyright   :  (c) Sergey Vinokurov 2019
+----------------------------------------------------------------------------
+
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE ConstraintKinds            #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MagicHash                  #-}
+{-# LANGUAGE MultiWayIf                 #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE UnboxedTuples              #-}
+
+{-# OPTIONS_GHC -O2 #-}
+
+module FastTags.LexerM
+    ( countInputSpace
+    , AlexInput(..)
+    , aiLineL
+    , AlexState(..)
+    , mkAlexState
+    , alexEnterBirdLiterateEnv
+    , alexEnterLiterateLatexEnv
+    , alexExitLiterateEnv
+    , pushContext
+    , modifyCommentDepth
+    , modifyQuasiquoterDepth
+    , modifyPreprocessorDepth
+    , takeText
+    , addIndentationSize
+    , checkQuasiQuoteEndPresent
+    , AlexM
+    , runAlexM
+    , alexSetInput
+    , alexSetNextCode
+    , alexInputPrevChar
+    , dropUntilNL
+    , dropUntilUnescapedNL
+    , dropUntilNLOr
+    , dropUntilNLOrEither
+    , alexGetByte
+    , unsafeTextHeadAscii
+    , unsafeTextHeadOfTailAscii
+    , unsafeTextHead
+    , utf8BS
+    , extractDefineOrLetName
+
+    , asCodeL
+    , asCommentDepthL
+    , asQuasiquoterDepthL
+    , asIndentationSizeL
+    , asPreprocessorDepthL
+    , asLiterateLocL
+    , asHaveQQEndL
+    ) where
+
+import Control.Applicative as A
+import Control.DeepSeq
+import Control.Monad.ST
+import Control.Monad.State.Strict
+import Control.Monad.Writer.Strict
+
+import Data.Char
+import Data.Int
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as TE
+import Data.Void (Void, vacuous)
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as C8
+import qualified Data.ByteString.Internal as BSI
+import qualified Data.Vector.Unboxed as U
+import qualified Data.Vector.Unboxed.Mutable as UM
+
+import Foreign.ForeignPtr
+import Foreign.Ptr
+import GHC.Base
+import GHC.Ptr
+import GHC.Word
+import Text.Printf
+
+import FastTags.LensBlaze
+import FastTags.LexerTypes
+import FastTags.Token
+
+#if __GLASGOW_HASKELL__ > 710
+import Data.Word (Word8)
+import GHC.IO (IO(..))
```

Comment:
Are you really using the IO constructor?  When I commented out this line it still compiled, so I guess not?

The Word8 above doesn't seem to be necessary either, probably because it's coming from GHC.Word

Reply:
I'm using IO because I'm defining my own version of `unafePerformIO` called `inlinePerformIO` (sprinkled throughout the module). Basically, it's like http://hackage.haskell.org/package/bytestring-0.10.10.0/docs/Data-ByteString-Internal.html#accursedUnutterablePerformIO. I think it might be a good idea to use that instead of defining my own, and thus resolve the comment.

Reply:
I see, I agree if we're going to use it, then using the one from bytestring is good because it already has documentation.  However, looking at the documentation, it's not exactly encouraging!  I would put on a comment saying why this, and why not unsafePerformIO, and why not unsafeInlinableIO (or if that's called unsafeDupableIO now, that one).  I see it's basically doing ptr manipulation.  I recall unsafePerformIO had some kind of tiny but present overhead and I've used unsafeDupableIO + alloca for a situation like that.  I assume you're using accursed for the same reason ByteString does... so maybe it would be enough to say that the rationale is the same, and see bytestring comments.